iroh-resolver/src/dns_resolver.rs

```diff
@@ -0,0 +1,157 @@
+}
+
+impl Config {
+    pub fn empty() -> Self {
+        Config {
+            tld_resolvers: None,
+        }
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Config {
+            /// Documentation on .eth TLD lives on https://eth.link/
+            tld_resolvers: Some(HashMap::from_iter(vec![(
+                "eth".to_string(),
+                ResolverConfig::from_parts(
+                    None,
+                    vec![],
+                    NameServerConfigGroup::from_ips_https(
+                        &[
+                            IpAddr::V4(Ipv4Addr::new(104, 18, 165, 219)),
+                            IpAddr::V4(Ipv4Addr::new(104, 18, 166, 219)),
+                        ],
+                        443,
+                        "resolver.cloudflare-eth.com".to_string(),
+                        true,
+                    ),
+                ),
+            )])),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct DnsResolver {
+    default_resolver: TokioAsyncResolver,
+    tld_resolvers: Option<HashMap<String, TokioAsyncResolver>>,
+}
+
+impl DnsResolver {
+    /// Creates resolver from its config
+    pub fn from_config(dns_resolver_config: Config) -> DnsResolver {
+        let tld_resolvers = dns_resolver_config
+            .tld_resolvers
+            .map(|dns_resolver_config| {
+                dns_resolver_config
+                    .into_iter()
+                    .map(|(tld, config)| {
+                        (
+                            tld,
+                            AsyncResolver::tokio(config, ResolverOpts::default()).unwrap(),
+                        )
+                    })
+                    .collect::<HashMap<_, _>>()
+            });
+        DnsResolver {
+            default_resolver: AsyncResolver::tokio(
+                ResolverConfig::default(),
+                ResolverOpts::default(),
+            )
+            .unwrap(),
+            tld_resolvers,
+        }
+    }
+
+    #[tracing::instrument]
+    pub async fn resolve_dnslink(&self, url: &str) -> Result<Vec<Path>> {
+        let url = format!("_dnslink.{}.", url);
+        let records = self.resolve_txt_record(&url).await?;
+        let records = records
+            .into_iter()
+            .filter(|r| r.starts_with("dnslink="))
+            .map(|r| {
+                let p = r.trim_start_matches("dnslink=").trim();
+                p.parse()
+            })
+            .collect::<Result<_>>()?;
+        Ok(records)
+    }
+
+    pub async fn resolve_txt_record(&self, url: &str) -> Result<Vec<String>> {
+        let tld = url.split('.').filter(|s| !s.is_empty()).last();
+        let resolver = tld
+            .and_then(|tld| {
+                self.tld_resolvers
+                    .as_ref()
+                    .and_then(|tld_resolvers| tld_resolvers.get(tld))
+            })
+            .unwrap_or(&self.default_resolver);
+        let txt_response = resolver.txt_lookup(url).await?;
```

Comment:
question: Currently if a non default resolver fails, we don't retry with the default resolver. Does that make the most sense, or should we try to use the default resolver if we fail with the specific one?

Reply:
Well, it depends on expected use-cases. I considered TLD-resolvers as ones that could deal with domains that default resolver couldn't deal, otherwise there is no even need to introduce this particular TLD-resolver.

Reply:
makes sense 👍